clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberState.java

```diff
@@ -48,26 +61,50 @@ public enum MemberState {
      * {@link org.apache.kafka.common.protocol.Errors#FENCED_MEMBER_EPOCH} error from the
      * broker. This is a recoverable state, where the member
      * gives up its partitions by invoking the user callbacks for onPartitionsLost, and then
-     * transitions to {@link #UNJOINED} to rejoin the group as a new member.
+     * transitions to {@link #JOINING} to rejoin the group as a new member.
      */
     FENCED,
 
     /**
-     * The member failed with an unrecoverable error
+     * The member transitions to this state when it is leaving the group after a call to
+     * unsubscribe. It stays in this state while releasing its assignment (calling user's callback
+     * for partitions revoked or lost), until the callback completes and a heartbeat request is
+     * sent out to effectively leave the group (without waiting for a response).
+     */
+    LEAVING_GROUP,
+
+    /**
+     * Member has completed releasing its assignment, and stays in this state until the next
+     * heartbeat request is sent out to leave the group.
      */
-    FAILED;
+    SENDING_LEAVE_REQUEST,
 
+    /**
+     * The member failed with an unrecoverable error.
+     */
+    FATAL;
+
+    /**
+     * Valid state transitions
+     */
     static {
-        // Valid state transitions
-        STABLE.previousValidStates = Arrays.asList(UNJOINED, RECONCILING);
 
-        RECONCILING.previousValidStates = Arrays.asList(STABLE, UNJOINED);
+        STABLE.previousValidStates = Arrays.asList(JOINING, ACKNOWLEDGING_RECONCILED_ASSIGNMENT);
+
+        RECONCILING.previousValidStates = Arrays.asList(STABLE, JOINING);
+
+        ACKNOWLEDGING_RECONCILED_ASSIGNMENT.previousValidStates = Arrays.asList(RECONCILING);
+
+        FATAL.previousValidStates = Arrays.asList(JOINING, STABLE, RECONCILING, ACKNOWLEDGING_RECONCILED_ASSIGNMENT);
```

Comment:
I'm surprised that the previous valid states for FATAL is not all of the other states, such as FENCED.

Reply:
That's OK. I was just asking an innocent question. Makes sense to me.